src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -183,56 +218,104 @@ protected override void Dispose(bool disposing)
             }
 
             // Create RequestCompletionSource object and save current values of handler settings.
-            RequestCompletionSource state = new RequestCompletionSource();
-            state.CancellationToken = cancellationToken;
-            state.RequestMessage = request;
-            state.Handler = this;
+            RequestCompletionSource state = new RequestCompletionSource
+            {
+                CancellationToken = cancellationToken,
+                RequestMessage = request,
+            };
 
-            Task.Factory.StartNew(
-                s =>
-                {
-                    var rcs = (RequestCompletionSource)s;
-                    rcs.Handler.StartRequest(rcs);
-                },
-                state,
-                CancellationToken.None,
-                TaskCreationOptions.DenyChildAttach,
-                TaskScheduler.Default);
+            BeginRequest(state);
 
             return state.Task;
         }
 
         #region Private methods
 
-        private void StartRequest(RequestCompletionSource state)
+        private async void BeginRequest(RequestCompletionSource state)
         {
-            HttpResponseMessage responseMessage = null;
-            SafeCurlHandle requestHandle = null;
+            SafeCurlHandle requestHandle = new SafeCurlHandle();
+            GCHandle stateHandle = new GCHandle();
+            bool needCleanup = false;
 
             if (state.CancellationToken.IsCancellationRequested)
             {
-                state.TrySetCanceled();
+                state.TrySetCanceled(state.CancellationToken);
```

Comment:
FWIW, I don't think this cancellation check is necessary.  We just checked for cancellation a few instructions earlier before calling BeginRequest.